deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator_test.go

```diff
@@ -88,3 +90,49 @@ func TestGenerateDeploy(t *testing.T) {
 		})
 	}
 }
+
+func TestGenerateSvcForCoordinator(t *testing.T) {
+	for _, tt := range []struct {
+		name          string
+		rss           *uniffleapi.RemoteShuffleService
+		serviceCntMap map[v1.ServiceType]int
+	}{
+		{
+			name: "with RPCNodePort",
+			rss:  buildRssWithLabels(),
+			serviceCntMap: map[v1.ServiceType]int{
+				"":                     2, // defaults to headless service
+				v1.ServiceTypeNodePort: 2,
+			},
+		},
+		{
+			name: "without rpcNodePort",
+			rss: func() *uniffleapi.RemoteShuffleService {
+				withoutRPCNodePortRss := buildRssWithLabels()
+				withoutRPCNodePortRss.Spec.Coordinator.RPCNodePort = make([]int32, 0)
+				withoutRPCNodePortRss.Spec.Coordinator.HTTPNodePort = make([]int32, 0)
+				return withoutRPCNodePortRss
+			}(),
+			serviceCntMap: map[v1.ServiceType]int{
+				"": 2,
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			assertion := assert.New(t)
+			_, _, services, _ := GenerateCoordinators(tt.rss)
+			result := make(map[v1.ServiceType]int)
+			for _, service := range services {
+				result[service.Spec.Type]++
```

Comment:
In a real cluster, headless service type is also ClusterIP, perhaps you can simply encapsulate a method to determine the svc type?